### ISSUE TITLE
[ENH]: Specifying figure size based on width and ratio/aspect

### ISSUE THREAD

**Issue by oscargus** [matplotlib/matplotlib#28758]
### Problem

While playing around with style files, I realize that for a paper it often makes sense to specify the width (set by the publisher) and ratio (determined by what we plot). Although one can of course read up and remember the figure width etc, it is quite convenient to just use a style file to determine the width and then set the ratio when creating the figure.

### Proposed solution

There is https://matplotlib.org/stable/api/_as_gen/matplotlib.figure.Figure.set_figheight.html so one way is of course to create the figure using the style file and then set the height suitably. 

However, some options are:

`set_figratio`, basically:
``` python
set_figheight(self.get_figwidth()*ratio)
```

Add a `figratio` keyword to Figure (and subplots and figure and ...) that takes the width, either from `figsize` or add a new rcParam for figwidth. Not sure which is less confusing...

(I think one can argue that the width should be kept when modifying the ratio based on the publisher-defined argument.)

Edit: searching a bit more I found https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.figaspect which does the exact opposite to what I want. Maybe an option is to come up with a good name for a similar function that determines the size with fixed width? Although the original idea was to make it "simple".

**Comment by rcomer** [comment]
> Edit: searching a bit more I found https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.figaspect which does the exact opposite to what I want. Maybe an option is to come up with a good name for a similar function that determines the size with fixed width? Although the original idea was to make it "simple".

Could we just add a parameter to `figaspect` so the user can choose whether the height or width is modified?

**Comment by oscargus** [comment]
Yeah, that would be a good way to deal with it! (Although I also hope that there can be a "simpler" way added to sort it out directly.)

**Comment by rcomer** [comment]
Ah, I missed that `figaspect` only returns the height and width and doesn't do anything to a figure.

**Comment by jklymak** [comment]
@timhoffm I hope we have vanishingly few users who can't do figure(figsize=(6, 6/2)) so I don't see the advantage of figure(figsize=figsize(6, aspect=2))

I think what @oscargus is after is the ability to se the default figsize in the style sheet, and then modify the height by an aspect_ratio kwarg so calls in the code would be figure(aspect=2).  

I think the only objection to that proposal would be that figure(figsize=(6, 6), aspect=2) would be ambiguous.  But if we were clear that the height would always be modified by the ratio, I think that could work. My gut reaction in such cases, however, is to just write my own wrapper around figure/subplots that does what I want.

**Comment by oscargus** [comment]
> I think what @oscargus is after is the ability to se the default figsize in the style sheet, and then modify the height by an aspect_ratio kwarg so calls in the code would be figure(aspect=2).

Correct! Something like:

``` python
import matplotlib.pyplot as plt
import matplotlib as mpl
import scienceplots

with mpl.style.context(["science", "ieee"]):
    fig, ax = plt.subplots(aspect=2)
```

with the rationale that it is often the width which is the parameter that should be set as "fixed" by the style file.

**Comment by timhoffm** [comment]
I see. Can we come up with a reasonable specification that can be passed via `figsize`? The problem I see with `aspect` is that it interacts with figsize `figure(figsize=(6, 6), aspect=2)` is logically inconsistent; on top of being ambiguous. Can we come up with an API that explicitly states "I want the figsize with this aspect via keeping width (or adjusting height)?

**Comment by jklymak** [comment]
So I am not good at fine-tuning in aspect space - much more comfortable just using inches.  So one possibility is to let `figsize=4` mean the figure is 4" high, and uses the default width.  The only slight confusion there is that the usual spec is (w, h) so having one element default to the second element is a bit strange.  However, I do think the "fixed width, variable height" case is common enough to warrant doing something here.

**Comment by timhoffm** [comment]
For only setting figure width supporting `figsize=(None, 4)` would be enough and more clear than `figsize=4`. We can easily do that, but I think the original request was asking for more.

**Comment by jklymak** [comment]
Oh `figsize=(None, 4)` would be good, and could be symmetric for folks who want `figsize=(6, None)`.    @oscargus may have a strong preference for an aspect ratio here, but as I said I don't personally think in terms of aspect ratios

**Comment by arkapudota** [comment]
I can work on this issue!

**Comment by oscargus** [comment]
> @oscargus may have a strong preference for an aspect ratio here, but as I said I don't personally think in terms of aspect ratios

I think that would do. Ideally, I'd use aspect (as it is easier to guessimate that the aspect should be 2 than to try different heights when you are not really sure about the exact width), but `(None, 4)` is still better than not being able to modify the height without knowing (bothering to figure out) the exact width (as you really do not want to change it).

**Comment by timhoffm** [comment]
Ok, let's allow figsize tuples with at most one of the two entries being `None`. @arkapudota do you want to work on this?

This does not restrict our ability to expand for a possible aspect definition later.

**Comment by arkapudota** [comment]
Yes I can work on this

**Comment by timhoffm** [comment]
I'll close this because it's difficult to define a reasonable API for aspects. Instead, let's focus on the simpler topic of partially setting width / height #31400.
